# My First Machine Learning Project

We will use a popular dataset used by most beginners, the iris dataset. Our goal is to predict what species of iris a certain flower is depending on 

### Load and split dataset

We will use train_test_split method to split the data into a test set and a training set


In [1]:
import pandas as pd
import matplotlib
import numpy as np
import IPython
import sklearn
from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split
from sklearn.neighbors import KNeighborsClassifier


iris = load_iris()
X_train, X_test, y_train, y_test = train_test_split(iris['data'], iris['target'],
random_state=0)

### Instantiate the KNearestNeighbors Class

Every machine learning model in scikit-learn is implemented in their own `Estimator` class.

We instantiate the class into an object. In this case the KNeighborsClassifier class has the number of neighbors as the parameter. We'll start by setting this to 1.

In [2]:

knn = KNeighborsClassifier(n_neighbors=1)

### Fit the training data

To build the model of the TRAINING SET, we call the fit method. The method takes  the training data(X_train) and training labels(y_train) as arguments.

In [11]:
knn.fit(X_train, y_train)


KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
           metric_params=None, n_jobs=1, n_neighbors=1, p=2,
           weights='uniform')

## Making Predictions

>"Imagine we found an iris in the wild with a sepal length of 5cm, a sepal width of 2.9cm, a petal length of 1cm anda petal width of 0.2cm. What species of iris would this be?"

We'll put this data into a numpy array like below:

In [9]:
X_new = np.array([[5,2.9,1,0.2]])
X_new.shape


(1, 4)

Then we'll try predicting using the knn object's predict method as follows:

In [10]:
prediction = knn.predict(X_new)
iris['target_names'][prediction]

array(['setosa'], dtype='<U10')

The KNN model predicts that our new iris is a Setosa species. The next step is to decide whether or not to trust this model!

### Evaluating the model

We can make predictions for irises in the test data and compare it against its label. We shall compute it's accuracy, the fraction of flowers  we get right

In [6]:
y_pred = knn.predict(X_test) #we get the labels from the predictions we get from knn predicting X_test
np.mean(y_pred == y_test) #how much of the predicted labels equal the actual test labels

0.9736842105263158

Alternatively, we may use the `score` method of the knn object

In [7]:
knn.score(X_test, y_test) # shows whether the X_test data predictions match the y_test labels 

0.9736842105263158

It seems our model should be expected to be correct 97% of the time for new irises

## Summary

We dealt with a problem where we needed to predict what species of iris a flower is based on the physical measurements of the flower. We have a correctly labeled data set to use. The dataset includes two numpy arrays. The array `X` is our data, and contains the features. The `y` array contains the class label (in this case its either 0,1, or 2)

We must split our dataset into a training set and a test set, randomly chosen, using the `train_test_split` method. We chose the K nearest neighbors classification algorith to make our predictions.

We instantiate the class which implements this algorithm, and build the model by using the `fit` method, passing `X_train` and `y_train` 

We evaluate the model's accuracy by using the `score` method.